README.org
* Project information

This is a fork of Xournal, the inking / PDF markup application written by
Denis Aroux. It is based on [[https://github.com/dmgerman/xournal][D. M. German's github branch]] and builds on
Xournal 0.4.5, adding several features. These include image pasting and
embedding, lasso selection tool, and autosave. Xournal provides subpixel
resolution, PDF annotation and export, and runs on both Linux and Windows. 

This branch uses patched 64-bit font tool components. When building on a
32-bit system, remove this patch via =patch -p1 -R < xournal-64bit.patch=

*Installation:*   see INSTALL or INSTALL.win32

*User's manual:*  see html-doc/manual.html

*Original project page:* [[http://sourceforge.net/projects/xournal]]

* Branches 

- master: the main branch with my code and many extra features
- upstream: original code maintained on Sourceforge
- debian:   ubuntu xournal packages

The latter two branches are inherited from dmgerman's repository.

* Changes by Leo Alekseyev:
  
** New features
   
- resize selection improvements:
  - fixed aspect ratio when resizing selection by dragging corner
  - larger margin for grabbing a corner handle
  - (cosmetic) added small selection box padding
- pressing stylus button-2 and dragging duplicates selection
- autosave patch ported to Windows (see below)
- image support (combination of my code and Simon Guest's code; see below)

** Patches applied

| Name                                | URL                                                                                 |
|-------------------------------------+-------------------------------------------------------------------------------------|
| Autosave (Timo Kluck)               | [[http://sourceforge.net/tracker/?func=detail&aid=3122782&group_id=163434&atid=827735][http://sourceforge.net/tracker/?func=detail&aid=3122782&group_id=163434&atid=827735]] |
| Part of enhance image (Simon Guest) | [[http://sourceforge.net/tracker/?func=detail&aid=2987520&group_id=163434&atid=827735][http://sourceforge.net/tracker/?func=detail&aid=2987520&group_id=163434&atid=827735]] |
| Lasso (Ian-Woo Kim)                 | [[http://sourceforge.net/tracker/?func=detail&aid=2679096&group_id=163434&atid=827735][http://sourceforge.net/tracker/?func=detail&aid=2679096&group_id=163434&atid=827735]] |


*** Image functionality

Images can be pasted from the clipboard; image copy/paste works between
Xournal instances. Images are embedded inside the XOJ files with Base64
encoding. They are currently stored at the original resolution, and are
resized as needed for display.  

The code is a combination of contributions by Leo Alekseyev and Simon Guest
(in particular SG's code handles PDF export of images).
Overall functionality is very close to that of SG's enhanced image patch on
Sourceforge, with the major difference that the images are appropriately
rescaled as the zoom level changes.

The code builds on the original insert image patch by Victor Saase

*** Autosave 

Original patch is Timo Kluck's modification of code from Edward Z. Yang and
Paulo Neves. The following changes were made to Timo Kluck's version:
- code made compatible with Leo Alekseyev's and D. M. German's current
  codebase
- autosaves are stored in the central location (default: ~/.xournal/autosaves/
  and ~/.xournal/autosaves_tmp/ for files that have not been saved). The
  current path of the file is encoded as a SHA1 hash appended to the file name.
- checking for possible autosaves is done on opening a file, not just program startup  
- autosave now works on Windows (PIDs are obtained from Windows process API (psapi.dll)).
  
*** Lasso

Original patch is due to Ian-Woo Kim. As written, it triggers a bug in
libgnomecanvas 2.30 that results in an invalid path and a subsequent segfault due
to poor error handling in libgnomecanvas.  A fixed version of
gnome_canvas_path_def_close_all function is included here.
  
Some of the code was refactored to avoid code duplication between rectangle
and lasso selection.

** Code reorganization 

- Separated out image, selection, and clipboard functionality into their
  own files (xo-image.c, xo-selection.c, xo-clipboard.c).

- Put include guards into all headers.

- Refactored much of the clipboard copy/paste code.

- Refactored rectangle selection code to maximize code reuse between
  lasso and rectangle selection functions.

** Build system 

- Modified automake files to enable seamless compilation under Windows
  (code was taken from Georg Wechslberger's branch at https://github.com/tauu/xournal)

- Added Windows installation instructions (Georg Wechslberger's modifications of
  Dirk Gerrits' instructions with some modifications by L.A.).

- Added 64-bit patch to the repository (originally found in the Xournal SUSE
  package). This patch only deals with the included Sun font tools (that
  code is not 64-bit safe).

* Patches from other sources
** Patches applied by D. M. German:

| Name                        | URL                                                                                           |
|-----------------------------+-----------------------------------------------------------------------------------------------|
| Last page seen              | https://sourceforge.net/tracker/index.php?func=detail&aid=3203323&group_id=163434&atid=827735 |
| Add next-prev file          | https://sourceforge.net/tracker/index.php?func=detail&aid=2730174&group_id=163434&atid=827735 |
| Cmd Line: Open a given page | https://sourceforge.net/tracker/index.php?func=detail&aid=2924825&group_id=163434&atid=827735 |
| Text font sizing            | https://sourceforge.net/tracker/index.php?func=detail&aid=3030878&group_id=163434&atid=827735 |
| Poor man's bookmarks        | https://sourceforge.net/tracker/index.php?func=detail&aid=2838282&group_id=163434&atid=827735 |
| Image insert                | http://sourceforge.net/tracker/?func=detail&aid=2890925&group_id=163434&atid=827735                                                                                              |

*** Load page seen

 Saves the current page number in the xoj file. When you open it
 again, jumps there

*** Add next-prev file

Opens the next/pref .xoj file  in the current directory

*** Text font sizing

Use Ctrl-B to bold, Ctrl > and Ctr < to increase size or reduce size
of annotation

*** Poor man's bookmarks

Jumps to "next notable" page, e.g. a page that has inking on it.  
This is mostly useful when working with marked-up PDF documents, which might
not have inking on every page.

*** Image insert

Basic image support from Victor Saase




ChangeLog
2010-02-12  dmg  <dmg@uvic.ca>

	* src/xo-callbacks.c (on_editRemember_activate): Forgot to release
	some memory

	* src/xo-interface.c, src/xo-callbacks.c (on_editRemember_activate): Added support for
	store, link

	* src/xo-misc.c (intercept_activate_events): Commented out a
	debugging string.

	* src/xo-callbacks.c (on_editRemember_activate): Added conversion
	of filenames to utf8.

	* src/xo-misc.c (encode_uri): Fixed bug with characters with first
	bit on.

Version 0.4.5 (Oct 2, 2009):
  - bugfixes for GTK+ 2.16/2.17 issues with xinput events
  - various minor UI bugfixes
  - gettext internationalization (contributed by David Planella)
  - Catalan translation (by David Planella), French translation
  - use poppler instead of pdftoppm to render PDF backgrounds 
    (after patches by Mike Ter Louw and Bob McElrath)
  - various improvements to UI and to key bindings (including
    patches by Bob McElrath and Lu Zhihe)
  - use gtk-print instead of libgnomeprint for printing
  - custom color chooser (patch contributed by Alex Ray)
  - option to have tablet buttons toggle the mapping rather than draw
  - paper color chooser (after a patch by Ole Joergen Broenner)
  - remove binary installer (due to binary incompatibilities)
  - UPDATED DEPENDENCIES: need gtk+ 2.10, poppler-glib 0.5.4

Version 0.4.2.1 (Mar 27, 2008):
  - bugfix for #1926757 (crash upon pasting variable-width stroke)
  - bugfix: set ruler/recognizer setting to default upon switching tools

Version 0.4.2 (Mar 25, 2008):
  - bugfixes for X.org 7.3; allow XInput and core events in reverse order
  - resize selection (patch contributed by Andy Neitzke)
  - pressure sensitive pen (variable stroke width) (patch by Andy Neitzke)
  - geometric shape recognizer (after an idea by Lukasz Kaiser) (see manual)
  - clean up compiler warnings (patch contributed by Danny Kukawka)

Version 0.4.1 (Sep 15, 2007):
  - bugfix: compatibility with new versions of pdftoppm (thanks to V. Ciancia)
  - GTK+ 2.11 event processing bugfix
  - minor bugfixes: hand tool, handling of filenames containing '&'
  - desktop and MIME files (thanks to Mathieu Bouchard) + updated installer
  - config options: left-handed scrollbar (contributed by Uwe Winter),
    hide some menu items (customizable in config file), auto-save preferences

Version 0.4.0.1 (Sep 3, 2007):
  - bugfixes for GTK+ 2.11 behavior (thanks to everyone who reported bugs)

Version 0.4 (Aug 15, 2007):
  - text tool (handles most TrueType and Type1 fonts)
  - font selection dialog and button
  - keyboard mappings (arrow keys)
  - menu mnemonics and shortcuts (suggestions by Jean-Baptiste Rouquier)
  - more responsive hand tool
  - bugfix for GTK+ 2.11 XInput behavior (thanks to Robert Gerlach)
  - various minor bugfixes and enhancements

Version 0.3.3 (Jan 31, 2007):
  - bugfix: upon loading a new file, zoom is set to default startup zoom
  - config option to allow input from a mouse or other core pointer device
  - config file entry to specify a default location for open/save
    (patch contributed by Andy Neitzke)
  - config file entries to customize visibility and position of toolbars
  - icon (thanks to Michele Codutti)

Version 0.3.2 (Nov 25, 2006):
  - preferences file and Save Preferences command
  - extra customization (via preferences file)
  - minor UI changes (patch contributed by Eduardo de Barros Lima)
  - hand tool (partially contributed by Vincenzo Ciancia)
  - a few bugfixes in rendering of bitmap backgrounds

Version 0.3.1 (Aug 3, 2006):
  - fixed a file format bug on systems with non-standard numeric locale
    (bug reported by Gert Renckens)

Version 0.3 (Jul 23, 2006):
  - new PDF rendering engine: export to PDF generates optimized files
    (smaller and more accurate)
  - export to PDF handles PDF backgrounds (up to PDF-1.4) natively
    (without conversion to bitmap)
  - default thickness of erasers and highlighters slightly increased
  - zoom dialog box with input box and "fit to page height" option
  - file format documentation added to the user's manual

Version 0.2.2 (Jun 5, 2006):
  - mapping of tools to stylus buttons (the options menu has new entries
    to allow the mapping of buttons 2 and 3 to arbitrary tools; the tools
    menu and toolbar affect the settings for button 1) (see manual)
  - moving selection by drag-and-drop works across page boundaries
  - vertical space tool can move items to next page (only when the entire
    block being moved has crossed the page boundary; items on the new page
    are not moved)
  - "apply to all pages" is now a toggle button affecting the behavior of
    paper size, color, and style commands
  - change in the behavior of the selection upon switching between tools

Version 0.2.1 (Jun 3, 2006):
  - recently used files listed in file menu
  - can change color or thickness of pen strokes in selection
  - function to apply paper style to all pages
  - can specify on command line a PDF file to annotate
  - suggest a derived file name for PDF annotation
  - speed up switching between pages
  - fixed a bug in XInput initialization (thanks to Luca de Cicco)
  - fixed a bug in print ranges (thanks to Mathieu Bouchard)
  - fixed a refresh bug in rescaling bitmap backgrounds

Version 0.2 (Jan 29, 2006):
  - PDF file annotation using xpdf's pdftoppm
     (PDF backgrounds updated asynchronously at all resolutions) 
  - PS/PDF backgrounds (as bitmaps at fixed resolution) using ghostscript
  - option to antialias and filter bitmap backgrounds when zooming
  - option to emulate eraser tip with right mouse button
  - binary installer allows non-root installation without compiling
  - full-screen mode (patch contributed by Luca De Cicco)

Version 0.1.1 (Jan 5, 2006): 
  - two bugfixes
  - backward compatibility with GTK+ 2.4

Version 0.1   (Jan 2, 2006): initial release


src/main.c
#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <sys/stat.h>
#include <string.h>
#include <gtk/gtk.h>
#include <libgnomecanvas/libgnomecanvas.h>

#include "xournal.h"
#include "xo-interface.h"
#include "xo-support.h"
#include "xo-callbacks.h"
#include "xo-misc.h"
#include "xo-file.h"
#include "xo-paint.h"
#include "xo-shapes.h"

static gint openAtPageNumber = 1;
static const char **fileArguments = NULL;


static gchar* exportPdfFile;
static gboolean show_gui = TRUE;

static GOptionEntry entries[] =
{
  { "page", 'p', 0, G_OPTION_ARG_INT, &openAtPageNumber, "Jump to Page", "N" },
  { "export-pdf", 'A', 0, G_OPTION_ARG_STRING, &exportPdfFile, "Export document to a PDF file", "FILENAME" },
  { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &fileArguments, NULL, N_("[FILE]") },
  { NULL }
};


GtkWidget *winMain;
GnomeCanvas *canvas;

struct Journal journal; // the journal
struct BgPdf bgpdf;  // the PDF loader stuff
struct UIData ui;   // the user interface data
struct UndoItem *undo, *redo; // the undo and redo stacks
  GError  *error = NULL;
  GOptionContext *context;

double DEFAULT_ZOOM;

void init_stuff (int argc, char *argv[])
{
  GtkWidget *w;
  GList *dev_list;
  GdkDevice *device;
  GdkScreen *screen;
  int i, j;
  struct Brush *b;
  gboolean can_xinput, success;
  gchar *tmppath, *tmpfn;
  GError  *error = NULL;
  GOptionContext *context;

  // create some data structures needed to populate the preferences
  ui.default_page.bg = g_new(struct Background, 1);

  // initialize config file names
  tmppath = g_build_filename(g_get_home_dir(), CONFIG_DIR, NULL);
  mkdir(tmppath, 0700); // safer (MRU data may be confidential)
  ui.mrufile = g_build_filename(tmppath, MRU_FILE, NULL);
  ui.configfile = g_build_filename(tmppath, CONFIG_FILE, NULL);
  g_free(tmppath);

  // initialize preferences
  init_config_default();
  load_config_from_file();
  ui.font_name = g_strdup(ui.default_font_name);
  ui.font_size = ui.default_font_size;
  ui.hiliter_alpha_mask = 0xffffff00 + (guint)(255*ui.hiliter_opacity);

  // we need an empty canvas prior to creating the journal structures
  canvas = GNOME_CANVAS (gnome_canvas_new_aa ());

  // initialize data
  ui.default_page.bg->canvas_item = NULL;
  ui.layerbox_length = 0;

  // parse command line options
  context = g_option_context_new ("");
  g_option_context_add_main_entries (context, entries, GETTEXT_PACKAGE);
  g_option_context_add_group (context, gtk_get_option_group (TRUE));
  if (!g_option_context_parse (context, &argc, &argv, &error)) {
      printf(_("option parsing failed: %s\n"), error->message);
      exit (1);
  }
  
  undo = NULL; redo = NULL;
  journal.pages = NULL;
  bgpdf.status = STATUS_NOT_INIT;

  new_journal();

  if(exportPdfFile) {
    show_gui = FALSE;
  }

  if (fileArguments == NULL) {
    if(exportPdfFile) {
       // If exporting having a file to open is essential
       printf(_("Path to xoj file mssing\n"));
       exit(2);
     }
  } else {
    if (g_path_is_absolute(fileArguments[0])) {  
      tmpfn = g_strdup(fileArguments[0]);
    } else {
      tmppath = g_get_current_dir();
      tmpfn = g_build_filename(tmppath, fileArguments[0], NULL);
      g_free(tmppath);
    }

    success = open_journal(tmpfn);
    g_free(tmpfn);
  }
  
  if (!success && !show_gui) { // GUI error is showed later
    printf(_("Error opening file '%s'"), argv[1]);
    exit(2);
  }

  if(exportPdfFile) {
     exit(print_to_pdf(exportPdfFile));
  }

  ui.cur_item_type = ITEM_NONE;
  ui.cur_item = NULL;
  ui.cur_path.coords = NULL;
  ui.cur_path_storage_alloc = 0;
  ui.cur_path.ref_count = 1;
  ui.cur_widths = NULL;
  ui.cur_widths_storage_alloc = 0;

  ui.selection = NULL;
  ui.cursor = NULL;

  ui.cur_brush = &(ui.brushes[0][ui.toolno[0]]);
  for (j=0; j<=NUM_BUTTONS; j++)
    for (i=0; i < NUM_STROKE_TOOLS; i++) {
      b = &(ui.brushes[j][i]);
      b->tool_type = i;
      if (b->color_no>=0) {
        b->color_rgba = predef_colors_rgba[b->color_no];
        if (i == TOOL_HIGHLIGHTER) {
          b->color_rgba &= ui.hiliter_alpha_mask;
        }
      }
      b->thickness = predef_thickness[i][b->thickness_no];
    }
  for (i=0; i<NUM_STROKE_TOOLS; i++)
    g_memmove(ui.default_brushes+i, &(ui.brushes[0][i]), sizeof(struct Brush));

  ui.cur_mapping = 0;
  ui.which_unswitch_button = 0;
  
  reset_recognizer();

  // initialize various interface elements
  
  gtk_window_set_default_size(GTK_WINDOW (winMain), ui.window_default_width, ui.window_default_height);
  if (ui.maximize_at_start) gtk_window_maximize(GTK_WINDOW (winMain));
  update_toolbar_and_menu();
  update_font_button();

  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("journalApplyAllPages")), ui.bg_apply_all_pages);
  if (ui.fullscreen) {
    gtk_check_menu_item_set_active(
      GTK_CHECK_MENU_ITEM(GET_COMPONENT("viewFullscreen")), TRUE);
    gtk_toggle_tool_button_set_active(
      GTK_TOGGLE_TOOL_BUTTON(GET_COMPONENT("buttonFullscreen")), TRUE);
    gtk_window_fullscreen(GTK_WINDOW(winMain));
  }
  gtk_button_set_relief(GTK_BUTTON(GET_COMPONENT("buttonColorChooser")), GTK_RELIEF_NONE);

  allow_all_accels();
  add_scroll_bindings();

  // prevent interface items from stealing focus
  // glade doesn't properly handle can_focus, so manually set it
  gtk_combo_box_set_focus_on_click(GTK_COMBO_BOX(GET_COMPONENT("comboLayer")), FALSE);
  g_signal_connect(GET_COMPONENT("spinPageNo"), "activate",
          G_CALLBACK(handle_activate_signal), NULL);
  gtk_container_forall(GTK_CONTAINER(winMain), unset_flags, (gpointer)GTK_CAN_FOCUS);
  GTK_WIDGET_SET_FLAGS(GTK_WIDGET(canvas), GTK_CAN_FOCUS);
  GTK_WIDGET_SET_FLAGS(GTK_WIDGET(GET_COMPONENT("spinPageNo")), GTK_CAN_FOCUS);
  
  set_flags(GET_COMPONENT("findBar"), GTK_CAN_FOCUS);

  // install hooks on button/key/activation events to make the spinPageNo lose focus
  gtk_container_forall(GTK_CONTAINER(winMain), install_focus_hooks, NULL);

  // set up and initialize the canvas

  gtk_widget_show (GTK_WIDGET (canvas));
  w = GET_COMPONENT("scrolledwindowMain");
  gtk_container_add (GTK_CONTAINER (w), GTK_WIDGET (canvas));
  gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW (w), GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);
  gtk_widget_set_events (GTK_WIDGET (canvas), GDK_EXPOSURE_MASK | GDK_POINTER_MOTION_MASK | GDK_BUTTON_MOTION_MASK | GDK_BUTTON_PRESS_MASK | GDK_BUTTON_RELEASE_MASK | GDK_KEY_PRESS_MASK | GDK_ENTER_NOTIFY_MASK | GDK_LEAVE_NOTIFY_MASK);
  gnome_canvas_set_pixels_per_unit (canvas, ui.zoom);
  gnome_canvas_set_center_scroll_region (canvas, TRUE);
  gtk_layout_get_hadjustment(GTK_LAYOUT (canvas))->step_increment = ui.scrollbar_step_increment;
  gtk_layout_get_vadjustment(GTK_LAYOUT (canvas))->step_increment = ui.scrollbar_step_increment;

  // set up the page size and canvas size
  update_page_stuff();

  g_signal_connect ((gpointer) canvas, "button_press_event",
                    G_CALLBACK (on_canvas_button_press_event),
                    NULL);
  g_signal_connect ((gpointer) canvas, "button_release_event",
                    G_CALLBACK (on_canvas_button_release_event),
                    NULL);
  g_signal_connect ((gpointer) canvas, "enter_notify_event",
                    G_CALLBACK (on_canvas_enter_notify_event),
                    NULL);
  g_signal_connect ((gpointer) canvas, "leave_notify_event",
                    G_CALLBACK (on_canvas_leave_notify_event),
                    NULL);
  g_signal_connect ((gpointer) canvas, "expose_event",
                    G_CALLBACK (on_canvas_expose_event),
                    NULL);
  g_signal_connect ((gpointer) canvas, "key_press_event",
                    G_CALLBACK (on_canvas_key_press_event),
                    NULL);
  g_signal_connect ((gpointer) canvas, "motion_notify_event",
                    G_CALLBACK (on_canvas_motion_notify_event),
                    NULL);
  g_signal_connect ((gpointer) gtk_layout_get_vadjustment(GTK_LAYOUT(canvas)),
                    "value-changed", G_CALLBACK (on_vscroll_changed),
                    NULL);
  g_object_set_data (G_OBJECT (winMain), "canvas", canvas);

  screen = gtk_widget_get_screen(winMain);
  ui.screen_width = gdk_screen_get_width(screen);
  ui.screen_height = gdk_screen_get_height(screen);
  
  can_xinput = FALSE;
  dev_list = gdk_devices_list();
  while (dev_list != NULL) {
    device = (GdkDevice *)dev_list->data;
    if (device != gdk_device_get_core_pointer()) {
      /* get around a GDK bug: map the valuator range CORRECTLY to [0,1] */
#ifdef ENABLE_XINPUT_BUGFIX
      gdk_device_set_axis_use(device, 0, GDK_AXIS_IGNORE);
      gdk_device_set_axis_use(device, 1, GDK_AXIS_IGNORE);
#endif
      gdk_device_set_mode(device, GDK_MODE_SCREEN);
      if (g_str_has_suffix(device->name, "eraser"))
        gdk_device_set_source(device, GDK_SOURCE_ERASER);
      can_xinput = TRUE;
    }
    dev_list = dev_list->next;
  }
  if (!can_xinput)
    gtk_widget_set_sensitive(GET_COMPONENT("optionsUseXInput"), FALSE);

  ui.use_xinput = ui.allow_xinput && can_xinput;

  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsProgressiveBG")), ui.progressive_bg);
  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsPrintRuling")), ui.print_ruling);
  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsAutoloadPdfXoj")), ui.autoload_pdf_xoj);
  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsLeftHanded")), ui.left_handed);
  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsShortenMenus")), ui.shorten_menus);
  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsAutoSavePrefs")), ui.auto_save_prefs);
  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsButtonSwitchMapping")), ui.button_switch_mapping);
  
  hide_unimplemented();

  update_undo_redo_enabled();
  update_copy_paste_enabled();
  update_vbox_order(ui.vertical_order[ui.fullscreen?1:0]);
  gtk_widget_grab_focus(GTK_WIDGET(canvas));

  // show everything...
  
  gtk_widget_show (winMain);
  update_cursor();

  /* this will cause extension events to get enabled/disabled, but
     we need the windows to be mapped first */
  gtk_check_menu_item_set_active(
    GTK_CHECK_MENU_ITEM(GET_COMPONENT("optionsUseXInput")), ui.use_xinput);

  /* fix a bug in GTK+ 2.16 and 2.17: scrollbars shouldn't get extended
     input events from pointer motion when cursor moves into main window */

  if (!gtk_check_version(2, 16, 0)) {
    g_signal_connect (
      GET_COMPONENT("menubar"),
      "event", G_CALLBACK (filter_extended_events),
      NULL);
    g_signal_connect (
      GET_COMPONENT("toolbarMain"),
      "event", G_CALLBACK (filter_extended_events),
      NULL);
    g_signal_connect (
      GET_COMPONENT("toolbarPen"),
      "event", G_CALLBACK (filter_extended_events),
      NULL);
    g_signal_connect (
      GET_COMPONENT("statusbar"),
      "event", G_CALLBACK (filter_extended_events),
      NULL);
    g_signal_connect (
      (gpointer)(gtk_scrolled_window_get_vscrollbar(GTK_SCROLLED_WINDOW(w))),
      "event", G_CALLBACK (filter_extended_events),
      NULL);
    g_signal_connect (
      (gpointer)(gtk_scrolled_window_get_hscrollbar(GTK_SCROLLED_WINDOW(w))),
      "event", G_CALLBACK (filter_extended_events),
      NULL);
  }

  // load the MRU
  
  init_mru();

  // here is the command line "option" is handled

  context = g_option_context_new ("Xournal");
  g_option_context_add_main_entries (context, entries, GETTEXT_PACKAGE);
  g_option_context_add_group (context, gtk_get_option_group (TRUE));
  if (!g_option_context_parse (context, &argc, &argv, &error)) {
    g_printerr ("Cannot parse arguments: %s\n", error->message);
    g_error_free (error);
    g_option_context_free (context);
    gtk_exit(0);
  }
  g_option_context_free (context);

  // show error if loading file failed
  
  // if no file was provided, we are done
  if (fileArguments == NULL || fileArguments[0] == NULL) return;
  //  fprintf(stderr, "Fiel name %s\n", fileArguments[0]);
  /*
  set_cursor_busy(TRUE);
  if (g_path_is_absolute(fileArguments[0]))
    tmpfn = g_strdup(fileArguments[0]);
  else {
    tmppath = g_get_current_dir();
    tmpfn = g_build_filename(tmppath, fileArguments[0], NULL);
    g_free(tmppath);
  }
  success = open_journal(tmpfn);
  g_free(tmpfn);
  set_cursor_busy(FALSE);
  */
  if (!success) {
	  #ifdef IMAGE_DEBUG
	  printf("error opening file '%s'\n",fileArguments[0]);
	  #endif
  
   // jump to desired page
   if (openAtPageNumber > journal.npages) {
     openAtPageNumber = journal.npages;
   }
   do_switch_page(openAtPageNumber - 1, TRUE, TRUE);
    w = gtk_message_dialog_new(GTK_WINDOW (winMain), GTK_DIALOG_DESTROY_WITH_PARENT,
       GTK_MESSAGE_ERROR, GTK_BUTTONS_OK, _("Error opening file '%s'"), fileArguments[0]);
    gtk_dialog_run(GTK_DIALOG(w));
    gtk_widget_destroy(w);
  }
  // jump to desired page, if requested...
  if (openAtPageNumber > 1) {
    //fprintf(stderr, "Jumping to page %d out of %d pages \n", openAtPageNumber, journal.npages);
    if (openAtPageNumber > journal.npages)
      // if user wants to jump too far away, then use last page
      openAtPageNumber = journal.npages;
    do_switch_page(openAtPageNumber - 1, TRUE, TRUE);
  }
}


int
main (int argc, char *argv[])
{
  gchar *path, *path1, *path2;
  
#ifdef ENABLE_NLS
  bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);
#endif
  
  gtk_set_locale ();
  gtk_init (&argc, &argv);

  add_pixmap_directory (PACKAGE_DATA_DIR "/" PACKAGE "/pixmaps");
  path = g_path_get_dirname(argv[0]);
  path1 = g_build_filename(path, "pixmaps", NULL);
  path2 = g_build_filename(path, "..", "pixmaps", NULL);
  add_pixmap_directory (path1);
  add_pixmap_directory (path2);
  add_pixmap_directory (path);
  g_free(path);
  g_free(path1);
  g_free(path2);

  /*
   * The following code was added by Glade to create one of each component
   * (except popup menus), just so that you see something after building
   * the project. Delete any components that you don't want shown initially.
   */
  winMain = create_winMain ();
  
  init_stuff (argc, argv);
  gtk_window_set_icon(GTK_WINDOW(winMain), create_pixbuf("xournal.png"));
  
  gtk_main ();
  
  if (bgpdf.status != STATUS_NOT_INIT) shutdown_bgpdf();

  save_mru_list();
  if (ui.auto_save_prefs) save_config_to_file();
  
  return 0;
}

